File: python/racs.py
# -*- coding: utf-8 -*-
#-------------------------------------------------------------------------------
'''опции командной строки:
  key=value --- задает значение параметра расчета key, конвертируется к типу 
значения параметра по умолчанию. Возможно задание серии значений параметра как 
  key=[expression]  --- expression вычисляется функцией eval в словаре модуля math
  key=[x1,x2..xn]   --- хаскелль-стиль для арифметической прогрессии
  key=[x1:step..xn] --- арифметическая прогрессия с шагом step
  key=[x1@step..xn] --- геометрическая прогрессия с множителем step
  key=[x1#size..xn] --- арифметическая прогрессия из size элементов
  key=[x1^size..xn] --- геометрическая прогрессия из size элементов
Значения x1, xn, step, size вычисляются функцией eval в словаре модуля math.
Параметры x1 и xn всегда включаются в серию. При явном задании параметра step шаг 
корректируется при необходимости для точного попадания в xn. Параметр size должен
быть целочисленным  (не менее двух). Если серии заданы для нескольких параметров,
вычисляются все возможные комбинации значений (декартово произведение).

Для всех параметров (кроме серийных) возможно дублирование, актуальным является 
последнее значение. Параметры командной строки перекрывают параметры из файла. 
Для булевых параметров имя параметра экивалентно заданию True. Длинные имена 
параметров могу задаваться как с одним, так и с двумя минусами, короткие имена
только с одним. Параметр path задается без минусов.

  -h, --help --- показать эту справку и выйти

  --repo=PATH --- задает путь к репозитрию для создания расчета
    path=PATH --- явно задает путь к директории расчета, если расчет сущестовал 
                  словарь расчета будет обновлен из директории расчета
  -p|--clean-path[=Y] --- очищать директория явно заданную расчета, при этом
                          словарь расчета не обновляется из директории
   -s|--symlink[=Y] --- создавать символическую ссылку '_' на последнюю 
                        директорию расчета
   -d|--daemonize[=N] --- "демонизировать" расчет при запуске (освободить терминал, 
                          вывод будет перенаправлен в logfile в директории расчета)
   -S|--statechecker[=Y] --- запускать демона, фиксирующего  в файле .RACS аварийное
                             завершение расчета
   -c|--copies[=1] --- число копий процесса при проведении расчетов с серийными 
                       параметрами
   -e|--on-exit[=Y] --- по завершениии расчета автоматически фиксировать время работы,
                        состояние расчета и сохранять измения параметров на диск
   -n|--calc-num[=3] --- число знаков в номере расчета (в текущем дне) при 
                         автоматичской генерации имени директории расчета 
   -P|--auto-pull[=Y] --- автоматически сохранять все параметры расчета из 
                          контролируемых расчетом объектов.

'''
#-------------------------------------------------------------------------------
import os, sys, math, gtable, calc
from calc import *
from math import *
#-------------------------------------------------------------------------------
def _make_path_hook(self):
    calc._make_path(self)
    if Calc._symlink and not os.path.samefile(os.getcwd(), Calc.repo):
        try:
            if os.path.islink('_') or os.path.exists('_'): os.remove('_')
            os.symlink(self.path, '_')
        except Exception, e: print>>sys.stderr, e
    if Calc._statechecker:
        import inspect
        f = open('/tmp/racs-schk-%i'%os.getpid(), 'w')
        print>>f, '''#!/usr/bin/python -S
# -*- coding: utf-8 -*-
import os, sys, cPickle, time, socket
d2s = lambda t: time.strftime('%Y.%%m.%d-%X', time.localtime(t))
'''
        print>>f, inspect.getsource(get_login)
        print>>f, inspect.getsource(mk_daemon)
        print>>f, inspect.getsource(set_output)
        print>>f, '''mk_daemon(); set_output()
while 1:
    if '%i' in os.listdir('/proc/'): time.sleep(10)
    else: 
        R = cPickle.load(open(%r+'.RACS'))
        if R['statelist'][-1][0]=='started':
            R['statelist'].append(('killed', get_login(), socket.gethostname(), d2s(time.time()), 'racs-statechecker'))
            cPickle.dump(R, open(%r+'.RACS', 'w'))
        break
        '''%(os.getpid(), self.path, self.path)
        f.close(); os.chmod(f.name, 0700); os.system(f.name); os.remove(f.name)
    if Calc._on_exit: atexit.register(_on_exit, self)
    if self._daemonize: mixt.set_output(self.path+'logfile')
    return self.path
#-------------------------------------------------------------------------------
def _init_hook(self):
    if 'path' in self.__dict__: 
        if os.path.exists(self.path) and Calc._clean_path: os.system('rm -rf %s/*'%self.path)
        if not os.path.exists(self.path): os.makedirs(self.path)
        if Calc._on_exit: atexit.register(_on_exit, self)
        if self._daemonize: mixt.set_output(self.path+'logfile')    
#-------------------------------------------------------------------------------
def _on_exit(calc): 
    import os, time, mixt, chrono
    try:
        runtime = chrono.Time(time.time()-self._starttime)
        print 'RUNTIME %s SIZE %s (.RACS %s) %s'%(runtime, os.popen('du -hs '+calc.path).readline().split()[0],
                                                  mixt.size2string(os.path.getsize(calc.path+'.RACS')), calc.path)
        # calc.update() ???
        if calc.statelist[-1][0]=='started':
            if hasattr(sys, 'last_value'): calc.add_state('stopped')
            else: calc.progress, calc.runtime = 1., runtime; calc.add_state('finished')
    finally: calc.commit()
#-------------------------------------------------------------------------------
if any(o in sys.argv[1:] for o in '-h -help --help'.split()):
    print __doc__
    print ''.join(l for l in open(sys.modules['__main__'].__file__) if '#@' in l),
    sys.exit()
#-------------------------------------------------------------------------------
Calc._starttime, Calc._make_path_hook, Calc._init_hook = time.time(), _make_path_hook, _init_hook
opts = { 'symlink':('s', True), 'daemonize':('d', False), 'statechecker':('S', True), 'repo':('', 'repo'),
         'on-exit':('e', True), 'calc-num':('n', 3), 'auto-pull':('P', True), 'clean-path':('p', True), 'copies':('c', 1) }
for k, v in opts.items(): setattr(Calc, k, v[1])
Calc._qargs, Calc._args, arg_seqs, arg_order, i = [], list(sys.argv[1:]), {}, [], 0
while i<len(Calc._args):
    A = Calc._args[i]
    if mixt.is_name_eq(A) and A.split('=', 1)[1][0]=='[' and A[-1]==']':
        arg, l = A.split('=', 1)
        try: L = eval(l, math.__dict__)
        except SyntaxError, e:
            s = l[1:-1]
            for t in '#^@:,':
                if s.count('..')==1 and s.split('..')[0].count(t)==1:
                    ab, c = s.split('..'); a, b = ab.split(t)
                    a, b, c = [float(eval(x, math.__dict__)) for x in (a, b, c)]
                    if t==',': b, t = b-a, ':'
                    if t==':': b, t = int((c-a)/b+1.5), '#'
                    if t=='#': d = (c-a)/(b-1); L = [a+d*j for j in range(int(b))]
                    if t=='@': b, t = int(log(c/a)/log(b)+1.5), '^' # c=a*d**(n-1)
                    if t=='^': d = exp(log(c/a)/(b-1)); L = [a*d**j for j in range(int(b))]
                    if not L: raise Exception('incorrect step or limits in expression '+A)
                    break
            else: raise Exception('incorrect sequence expression '+A)
        arg_seqs[arg] = L; arg_order.append(arg); del Calc._args[i]
    elif mixt.is_name_eq(A): Calc._qargs.append(tuple(A.split('=', 1))); del Calc._args[i]
    elif A.startswith('-') and A!='-':
        for k, v in opts.items():
            if type(v[1]) is bool and any(A==x for x in ('-'+k, '--'+k, '-'+v[0])):
                setattr(Calc, k, True); del Calc._args[i]; break
            elif any(A.startswith(x+'=') and not A.split('=', 1)[1].startswith('=') for x in ('-'+k, '--'+k, '-'+v[0])):
                x = A.split('=', 1)[1]
                setattr(Calc, k, mixt.string2bool(x) if type(v[1]) is bool else int(x) if type(v[1]) is int else x)
                del Calc._args[i]; break
        else: i += 1
    else: i += 1
#-------------------------------------------------------------------------------
if Calc._daemonize: mixt.mk_daemon()
#-------------------------------------------------------------------------------
if arg_seqs:
    a0, i, copies, pids = arg_order.pop(0), 1, 1, []
    queue = reduce(lambda L, a: [l+[(a,x)] for x in arg_seqs[a] for l in L], arg_order, [[(a0,x)] for x in arg_seqs[a0]])
    print 'Start queue for %i items in %i threads, master PID=%i'(len(queue), copies, os.getpid())
    if Calc._daemonize: set_output()
    for q in queue:
        if len(pids)==copies:
            p = os.waitpid(-1, 0)[0]
            pids.remove(p)
        Calc._qargs += q+[('master', os.getpid())]
        pid = os.fork()
        if not pid: break
        pids.append(pid)
    else:
        while(pids): pids.remove(os.waitpid(-1, 0)[0])
        sys.exit()
#-------------------------------------------------------------------------------
__all__ = ['Calc']

File: src/racs.cpp
/**
 * Copyright (C) 2017 Antov V. Ivanov  <aiv.racs@gmail.com>
 * Licensed under the Apache License, Version 2.0
 **/

//#include <mpi.h>
#include <omp.h>
#include <stdlib.h>
#include <unistd.h>
#include <vector>
#include <list>
#include <wait.h>
#include <utime.h>
#include <ctime>
#include <fstream>
#include "../include/aiwlib/configfile"
#include "../include/aiwlib/mixt"
#include "../include/aiwlib/racs"
#include "../include/aiwlib/mpi4py"
using namespace aiw;

const char *racs_help = "опции командной строки:\n"
	"  key=value --- задает значение параметра расчета key, возможно задание серии значений параметра как\n"
	"  key=[x1,x2,...,xn]  --- перечисление значений через запятую\n"
	"  key=[x1,x2..xn]     --- хаскелль-стиль для арифметической прогрессии\n"
	"  key=[x1:step..xn] --- арифметическая прогрессия с шагом step\n"
	"  key=[x1@step..xn] --- геометрическая прогрессия с множителем step\n"
	"  key=[x1#size..xn] --- арифметическая прогрессия из size элементов\n"
	"  key=[x1^size..xn] --- геометрическая прогрессия из size элементов\n"
	"Значения x1, xn, step приводятся к типу double. Параметры x1 и xn\n"
	"всегда включаются в серию. При явном задании параметра step шаг всегда\n"
	"корректируется для точного попадания в xn. Параметр size должен быть\n"
	"целочисленным  (не менее двух). Если серии заданы для нескольких\n" 
	"параметров, вычисляются все возможные комбинации значений (декартово произведение).\n"

	"\n  -h|--help --- показать эту справку и выйти\n\n"

	"Для всех параметров (кроме серийных) возможно дублирование, актуальным является\n"
	"последнее значение. Для булевых параметров имя параметра экивалентно заданию значения True, кроме того\n"
	"можно использовать значения  Y|y|YES|Yes|yes|ON|On|on|TRUE|True|true|V|v|1 или\n"
	"N|n|NO|No|no|OFF|Off|off|FALSE|False|false|X|x|0.\n\n"

	"  -r|--repo=PATH --- задает путь к репозитрию для создания расчета\n"
    "  path=PATH --- явно задает путь к директории расчета\n" 
	"  -p|--clean-path[=Y] --- очищать явно заданную директорию расчета\n"
	"  -s|--symlink[=Y] --- создавать символическую ссылку '_' на последнюю\n" 
	"                       директорию расчета\n"
	"  -S|--statechecker[=Y] --- запускать демона, фиксирующего  в файле .RACS аварийное\n"
	"                            завершение расчета\n"
	"  -c|--copies=1 --- число копий процесса при проведении расчетов с серийными\n"
	"                    параметрами\n"
	"  --mpi[=N] --- для запуска из под MPI.\n\n";
//------------------------------------------------------------------------------
bool startswith(const std::string &str, const char *substr){
	for(size_t i=0; substr[i]; i++) if(i==str.size() || substr[i]!=str[i]) return false;
	return true;
}
bool parse_opt(const std::string& arg, std::string patt, bool &opt){
	if(startswith(arg, (patt+"=").c_str())){ std::stringstream buf(arg.substr(patt.size()+1, arg.size())); scanf_obj(buf, opt); return true; }
	else if(startswith(arg, patt.c_str())){ opt = true; return true; }
	return false;
}
template <typename T> bool parse_opt(const std::string& arg, std::string patt, T &opt){
	if(startswith(arg, patt.c_str())){ std::stringstream buf(arg.substr(patt.size(), arg.size())); scanf_obj(buf, opt); return true; }
	return false;
}
void set_queue_item(std::vector<int> &qpos, const std::vector<std::string> &qkeys,
					const std::vector<std::vector<double> >& qgrid, std::map<std::string, std::string> &arg_params){
	for(size_t i=0; i<qpos.size(); ++i){
		std::stringstream buf; buf<<qgrid[i][qpos[i]];
		arg_params[qkeys[i]] = buf.str();
	} 
}
bool next_queue_item(std::vector<int> &qpos, const std::vector<std::vector<double> >& qgrid){
	//if(qpos.back()>=int(qgrid.back().size())) return false;
	qpos[0]++; 
	for(int i=0; i<int(qpos.size())-1; i++)
		if(qpos[i]<int(qgrid[i].size())) break;
		else{ qpos[i] = 0; qpos[i+1]++; }
	return qpos.back()<int(qgrid.back().size());
}
//------------------------------------------------------------------------------
aiw::RacsCalc::RacsCalc(int argc, const char **argv){
	bool clean_path = true, use_mpi = false, schecker=true;
	int copies = 1;
	std::map<std::string, std::vector<double> > qparams;
	std::list<std::string> qparams_order;
	for(int i=0; i<argc; i++) args.push_back(argv[i]);
	for(int i=1; i<argc; i++){
		std::string a = argv[i];
		if(a=="-h" || a=="--help"){ std::cout<<racs_help; exit(0); }
		if(a.back()=='+'){ a.resize(a.size()-1); tags.push_back(a); continue; }
		if(parse_opt(a, "-r=", repo) || parse_opt(a, "--repo=", repo) || parse_opt(a, "path=", path_) ||
		   parse_opt(a, "-p", clean_path) || parse_opt(a, "--clean-path", clean_path) ||
		   parse_opt(a, "-s", this->symlink) || parse_opt(a, "--symlink", this->symlink) ||
		   parse_opt(a, "-S", schecker) || parse_opt(a, "--statechecker", schecker) ||
		   parse_opt(a, "-c=", copies) || parse_opt(a, "--copies=", copies) ||
		   parse_opt(a, "--mpi", use_mpi)) continue;
		size_t eq_pos = 0; while(eq_pos<a.size() && a[eq_pos]!='=') ++eq_pos;
		if(eq_pos<a.size()){
			std::string key = a.substr(0, eq_pos), val = a.substr(eq_pos+1, a.size());
			if(val.front()=='[' and val.back()==']'){ // подготовка очереди
				std::vector<double> &L = qparams[key]; qparams_order.push_back(key);
				if(arg_params.find(key)!=arg_params.end()) arg_params.erase(key);
				std::stringstream buf(val.substr(1, val.size()-2));
				size_t dd = val.find("..");
				if(dd<val.size()){
					double a, b, c=::atof(val.substr(dd+2, val.size()-dd-3).c_str()); char t; buf>>a>>t>>b; // проверить состояние ошибки buf!!!
					if(t==','){ b -= a; t = ':'; }
                    if(t==':'){ b = int((c-a)/b+1.5); t = '#'; }
                    if(t=='#'){ double d = (c-a)/(b-1); for(int j=0; j<int(b); j++) L.push_back(a+d*j); }
                    if(t=='@'){ b = int(log(c/a)/log(b)+1.5); t = '^'; } // c=a*d**(n-1)
                    if(t=='^'){ double d = exp(log(c/a)/(b-1)); for(int j=0; j<int(b); j++){ L.push_back(a); a *= d; } }
					if(L.empty()) WRAISE("incorrect step or limits in expression: ", a, key, val, a, b, c, t);
				} else do { double x; buf>>x; L.push_back(x); } while(buf.get()==','); //???
			} else{
				arg_params[key] = val;
				if(qparams.find(key)!=qparams.end()) qparams.erase(key);
				auto I=qparams_order.begin(); while(I!=qparams_order.end()) if(*I==key) qparams_order.erase(I++); else ++I;
			}
		}
	}
	if(qparams.empty()){ // нет пакетного запуска
		if(!path_.empty() && clean_path && system(("rm -rf "+path_+"/*").c_str()));
		if(!path_.empty()) calc_configure();
	} else { // есть пакетный запуск
		std::vector<std::string> qkeys(qparams.size());          // имена праметров
		std::vector<std::vector<double> > qgrid(qparams.size()); // сетка значений параметров
		int len_queue=1, n_start=0, n_finish=0; 
		{ int i=0; for(auto K=qparams_order.begin(); K!=qparams_order.end(); ++K){
				auto I = qparams.find(*K); len_queue *= I->second.size();
				qkeys[i] = I->first; qgrid[i++].swap(I->second);
			}
		}
		std::vector<int> qpos(qparams.size(), 0);  // позиция в сетке значений для пакетного запуска
		if(use_mpi){
			MPI_Init(&argc, (char***)&argv);
			int procID = mpi_proc_number(), proc_count = mpi_proc_count();
			if(procID==0){ // головной процесс
				std::ostream *logs[2]; out_preambule(logs, "", argc, argv, proc_count-1, qkeys, qgrid);
				schecker = clean_path = false;
				do { // цикл по заданиям
					aiw_mpi_msg_t msg; mpi_recv(msg, -1); // получаем запрос на задание, но на самом деле нам интересен только процесс от которого он пришел
					std::stringstream sbuf; for(size_t i=0; i<qpos.size(); ++i) sbuf<<qpos[i]<<' ';
					set_queue_item(qpos, qkeys, qgrid, arg_params); 				  
					path_ = ""; sbuf<<path(); mpi_send(sbuf.str(), msg.source); // формируем задание и отправляем его вычислителю
					out_start_task(logs, msg.source, double(++n_start)/len_queue, qpos, qkeys, qgrid);
				} while(next_queue_item(qpos, qgrid)); // конец цикла по заданиям
				for(int i=1; i<proc_count; i++) mpi_send("bye", i); // рассылаем сообщение о завершении работы
				out_close_log(logs, "");
			} else { // вычислитель
				char hostname[256]; ::gethostname(hostname, 255); 
				while(1){
					mpi_send(hostname, 0); // посылаем запрос на задание
					aiw_mpi_msg_t msg; mpi_recv(msg, 0); // получаем задание
					if(msg.data=="bye") break;
					if(fork()){	int wstatus;  waitpid(-1, &wstatus, 0); } // головной процесс
					else{ // дочерний процесс
						std::stringstream buf(msg.data);
						for(int i=0; i<int(qpos.size()); i++) buf>>qpos[i];
						path_ = buf.str().substr(int(buf.tellg())+1, 4095);
						set_queue_item(qpos, qkeys, qgrid, arg_params);  				  
						set_state("started");
						return;
					}
				}
			}
			MPI_Finalize();
			exit(0);
		} else { // not MPI
			std::ostream *logs[2]; out_preambule(logs, "/tmp/", argc, argv, copies,  qkeys, qgrid);
			std::list<pid_t> pids; int wstatus;
			do{
				if(int(pids.size())==copies){
					pid_t pid = waitpid(-1, &wstatus, 0);
					pids.remove(pid); out_finish_task(logs, pid, double(++n_finish)/len_queue);
				}
				pid_t pid = fork();
				if(!pid){ set_queue_item(qpos, qkeys, qgrid, arg_params); set_state("started"); return; } // дочерний процесс
				pids.push_back(pid); out_start_task(logs, pid, double(++n_start)/len_queue, qpos,  qkeys, qgrid);
			} while(next_queue_item(qpos, qgrid));
			while(pids.size()){ pid_t pid = waitpid(-1, &wstatus, 0); pids.remove(pid); out_finish_task(logs, pid, double(++n_finish)/len_queue); }
			out_close_log(logs, "/tmp/"); exit(0);
		}
	}
	set_state("started");
}
//------------------------------------------------------------------------------
aiw::RacsCalc::~RacsCalc(){
	if(!path_.empty()){
		// runtime = omp_get_wtime()-starttime;
		if(std::uncaught_exception()) set_state("stopped");
		else{ progress = 1.; set_state("finished"); }
		commit();
	}
}
//------------------------------------------------------------------------------
const std::string& aiw::RacsCalc::path(){
	if(path_.empty()){ // создаем уникальную директорию расчета
		std::string frepo = format_string(repo.c_str(), arg_params); // ???
		path_ = make_path(frepo.c_str());
		if(path_.empty()) WRAISE("cann't make unique path for calc: ", repo, frepo);
		calc_configure();
	}
	return path_;
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::calc_configure(){ // конфигурация расчета когда известен путь - настраиваем демона для завершения и т.д.
	if(this->symlink){ ::unlink("_"); if(::symlink(path().c_str(), "_")) WARNING("can't create symlink '_': ", path(), strerror(errno)); }
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::set_state(const std::string& state){
	if(state=="started") starttime = omp_get_wtime();
	state_t st; st.state = state; st.date = std::time(nullptr);
	statelist.push_back(st);
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::push(){ // push parametrs to controls
	for(auto I=controls.begin(); I!=controls.end(); ++I) (*I)->update(*this);
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::commit(){
	Pickle P = pickle_dict();	
	P("progress", progress)("runtime", pickle_class("aiwlib.chrono", "Time", false)<<runtime)("md5sum", md5sum);	
	{ Pickle L = pickle_list(); for(auto I=args.begin(); I!=args.end(); ++I) L<<*I; P("args", L); }	
	{ Pickle L = pickle_set();  for(auto I=tags.begin(); I!=tags.end(); ++I) L<<*I; P("tags", L); }	
	{ Pickle L = pickle_list();
		char hostname[256]; ::gethostname(hostname, 255);
		for(auto I=statelist.begin(); I!=statelist.end(); ++I){
			L<<pickle_tuple(I->state, getenv("USER"), hostname,	pickle_class("aiwlib.chrono", "Date", false)<<I->date, getpid());
		}
		P("statelist", L);
	}		
	for(auto I=controls.begin(); I!=controls.end(); ++I) (*I)->pickle(P);
	for(auto I=params.begin(); I!=params.end(); ++I) I->second->pickle(I->first, P);	
	std::ofstream(path()+".RACS")<<P.buf.str()<<'.';
	utime(path_.c_str(), nullptr);
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::set_progess(double progress_){
	progress = progress_; runtime = omp_get_wtime()-starttime; 
	commit();
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::out_preambule(std::ostream **logs, const char *logdir, int argc, const char **argv, int copies,
								  const std::vector<std::string> &qkeys, const std::vector<std::vector<double> > &qgrid){
	int len_queue = 1; for(auto I=qgrid.begin(); I!=qgrid.end(); ++I) len_queue *= I->size();
	char hostname[256], logfile[256]; ::gethostname(hostname, 255); snprintf(logfile, 255, "%sracs-started-%i", logdir, getpid());
	logs[0] = &std::cout; logs[1] = new	std::ofstream(logfile);
	printf("Start queue for %i items in %i threads, master PID=%i, logfile=\"%s\"", len_queue, copies, getpid(), logfile);
	for(int k=0; k<2; k++){
		(*logs[k])<<"# "<<getenv("USER")<<'@'<<hostname<<':'<<get_current_dir_name()<<" repo="<<repo
				  <<" tasks="<<len_queue<<" threads="<<copies<<"\n#";
		for(int i=0; i<argc; i++){ (*logs[k])<<' '<<argv[i]; } (*logs[k])<<'\n';
		for(size_t i=0; i<qkeys.size(); i++){
			(*logs[k])<<"#   "<<qkeys[i]<<": ["<<qgrid[i][0];
			for(size_t j=0; j<qgrid[i].size(); ++j) (*logs[k])<<", "<<qgrid[i][j];
			(*logs[k])<<"]\n";
		}
		logs[k]->flush();
	}
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::out_start_task(std::ostream **logs, int PID, double sprogress, const std::vector<int> &qpos,
								   const std::vector<std::string> &qkeys, const std::vector<std::vector<double> > &qgrid){
	for(int k=0; k<2; k++){
		*logs[k]<<date2string(time(nullptr))<<" +"<<PID<<' '<<sprogress*100<<'%';
		for(size_t i=0; i<qpos.size(); ++i) *logs[k]<<' '<<qkeys[i]<<'='<<qgrid[i][qpos[i]];
		*logs[k]<<'\n'; logs[k]->flush();
	}
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::out_finish_task(std::ostream **logs, int PID, double sprogress){
	double rt = omp_get_wtime()-starttime;
	for(int k=0; k<2; k++){
		*logs[k]<<date2string(time(nullptr))<<" -"<<PID<<' '<<sprogress*100<<"% "<<time2string(rt)<<" "<<time2string(rt/sprogress)<<'\n';
		logs[k]->flush();
	}
}
//------------------------------------------------------------------------------
void aiw::RacsCalc::out_close_log(std::ostream **logs, const char *logdir){
	delete logs[1];
	char logfile1[256], logfile2[256]; int pid =  getpid();
	snprintf(logfile1, 255, "%sracs-started-%i", logdir, pid);
	snprintf(logfile2, 255, "%sracs-finished-%i", logdir, pid);
	rename(logfile1, logfile2);
}
//------------------------------------------------------------------------------
